### PR TITLE
Implement ufm subnet manager client

### PR DESCRIPTION
**Pull request by Mmduh-483**

### CODE REVIEW

File: pkg/sm/plugins/ufm/ufm.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	req.SetBasicAuth(u.conf.Username, u.conf.Password)
+	req.Header.Set("Content-Type", "application/json; charset=utf-8")
+	return req, nil
+}
+
+func (u *ufmPlugin) createUfmUrl(path string) string {
+	return fmt.Sprintf("%s://%s:%d%s", u.conf.HttpSchema, u.conf.Address, u.conf.Port, path)
+}
+
+func (u *ufmPlugin) executeRequest(method, path string, expectedStatusCode int, body []byte) error {
```
Comment (moshe010) [inline pkg/sm/plugins/ufm/ufm.go]: I prefer that you will use u.client.Get/u.client.Post/u.client.Delete and remove this method

File: pkg/sm/plugins/ufm/ufm.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	u.client = client
+
+	if err := u.executeRequest("GET", "/ufmRest/app/ufm_version", 200, nil); err != nil {
+		return fmt.Errorf("failed to connect to fum subnet manger %v", err)
+	}
+
+	return nil
+}
+
+func (u *ufmPlugin) AddPKey(guid, pKey string) error {
```
Comment (moshe010) [inline pkg/sm/plugins/ufm/ufm.go]: I think we should do bulk call for UFM instead of call per container creation.

File: pkg/sm/plugins/ufm/ufm.go (hunk 1 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: Changed

File: pkg/sm/plugins/ufm/ufm.go (hunk 2 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: Changed

File: pkg/sm/plugins/plugin.go [hunk 3]
```diff
@@ -4,14 +4,17 @@ type SubnetManagerClient interface {
...
 
-	// Connect Check the client can reach the subnet manager and return error in case if it is not reachable.
+	// Validate Check the client can reach the subnet manager and return error in case if it is not reachable.
 	Validate() error
 
-	// AddPKey add pkey for the given guid.
-	// It return error if failed to add the pkey for specified guid.
-	AddPKey(pkey, guid string) error
+	// Config reads the connection conf to connect to subnet manager
+	Config(conf []byte) error
```
Comment (adrianchiris) [inline pkg/sm/plugins/plugin.go]: for the plugin framework you could:

1. have a mapping between plugin and conf file path per plugin (given in the global daemon config)
2. Config() here will get a path to the plugin specific configuration file.

File: pkg/sm/plugins/plugin.go [hunk 4]
```diff
@@ -4,14 +4,17 @@ type SubnetManagerClient interface {
...
 
-	// RemovePKey remove pkey for the given guid.
-	// It return error if failed to remove the pkey for specified guid.
-	RemovePKey(pkey, guid string) error
+	// AddGuidsToPKey add pkey for the given guid.
+	// It return error if failed.
+	AddGuidsToPKey(pkey string, guids []string) error
```
Comment (adrianchiris) [inline pkg/sm/plugins/plugin.go]: maybe guids type : []net.HardwareAddr ?

File: pkg/sm/plugins/plugin.go [hunk 5]
```diff
@@ -4,14 +4,17 @@ type SubnetManagerClient interface {
...
+
+	// RemoveGuidsFromPKey remove guids for given pkey.
+	// It return error if failed.
+	RemoveGuidsFromPKey(pkey string, guids []string) error
```
Comment (adrianchiris) [inline pkg/sm/plugins/plugin.go]: same as above

File: pkg/sm/plugins/ufm/main/main.go [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package main
+
+import (
+	"github.com/Mellanox/ib-kubernetes/pkg/sm/plugins"
+	"github.com/Mellanox/ib-kubernetes/pkg/sm/plugins/ufm"
+
+	"github.com/golang/glog"
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/main/main.go]: imports order should be : builtin, 3rd party, project specific

File: pkg/sm/plugins/ufm/ufm.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/Mellanox/ib-kubernetes/pkg/sm/plugins"
+
+	"github.com/golang/glog"
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/ufm.go]: import order

File: pkg/sm/plugins/ufm/ufm.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+type httpClient struct {
+	conf   *config
+	client http.Client
+}
+
+func (c *httpClient) Validate() error {
+	glog.V(3).Info("http-client Validate():")
+	// check required fields
+	if c.conf.Username == "" || c.conf.Password == "" || c.conf.Address == "" {
+		return fmt.Errorf(`failed to find one or more required filed of ["username", "password", "address"]`)
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/ufm.go]: /s/filed/field/

File: pkg/sm/plugins/ufm/ufm.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+	}
+
+	return nil
+}
+
+func (c *httpClient) Get(path string, expectedStatusCode int) error {
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/ufm.go]: can we be consistent with
http.Client.Get()
http.Client.Post()
methods signature for the Get/Post defined here?

later on this could be extended and moved to its own pkg

File: pkg/sm/plugins/ufm/ufm.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+	"net/http"
+	"strings"
+
+	"github.com/Mellanox/ib-kubernetes/pkg/sm/plugins"
+
+	"github.com/golang/glog"
+)
+
+type config struct {
+	Username   string `json:"username"`
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/ufm.go]: UFM supports certificates right ?

maybe add an AuthMode field ?
initial value can be "Basic"

File: pkg/sm/plugins/ufm/ufm.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+	return nil
+}
+
+type ufmPlugin struct {
+	client *httpClient
+}
+
+func NewUfmPlugin() plugins.SubnetManagerClient {
+	return &ufmPlugin{}
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/ufm.go]: who allocates and initializes client ?

File: pkg/sm/plugins/ufm/ufm.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+func (u *ufmPlugin) Name() string {
+	return "ufm"
+}
+
+func (u *ufmPlugin) Config(conf []byte) error {
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/ufm.go]: see my general comment about Config()

File: pkg/sm/plugins/ufm/ufm.go (hunk 11 above)
Comment (adrianchiris) [reply pkg/sm/plugins/ufm/ufm.go]: you are returning an object that if Config() is not called and finishes successfully will yield an unusable object that will cause the program to panic

File: pkg/sm/plugins/ufm/ufm.go [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+type config struct {
+	Username   string `json:"username"`
+	Password   string `json:"password"`
+	Address    string `json:"address"`
+	Port       int    `json:"port"`
+	HttpSchema string `json:"httpSchema"`
+}
+
+type httpClient struct {
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/ufm.go]: i think this should be in its own package

File: pkg/sm/plugins/plugin.go (hunk 4 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin.go]: ufm will receive a string not net.HardwareAddr, the plugin doesn't need to do the formatting, also it is not needed for the logic here.

File: pkg/sm/plugins/plugin.go (hunk 5 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin.go]: same as above

File: pkg/sm/plugins/plugin.go (hunk 3 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin.go]: The daemon reads the secret from the k8s then pass it to the plugin.

File: pkg/sm/plugins/plugin.go (hunk 4 above)
Comment (adrianchiris) [reply pkg/sm/plugins/plugin.go]: having it as HardwareAddr makes a stronger contract on the guid format.
you are defining the interface here, it should not depend on ufm.

File: pkg/sm/plugins/ufm/ufm.go (hunk 9 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: added

File: pkg/sm/plugins/plugin.go (hunk 3 above)
Comment (adrianchiris) [reply pkg/sm/plugins/plugin.go]: the secret is mounted in the container no ?
if an attacker gains access to the container, he could extract the user/password regardless if its in memory or file. 

see https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets

File: pkg/sm/plugins/ufm/ufm.go (hunk 11 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: the daemon will initialize the plugin client.
There is 2 ways: 
1- the current one
2- let the plugin client communicate with k8s rest api then read the secret?
@moshe010 @adrianchiris  which approach suite better?

File: pkg/sm/plugins/ufm/ufm.go (hunk 13 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: it won't be used any where else

File: pkg/sm/plugins/ufm/ufm.go (hunk 10 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: We can support the certificates in later PR

Comment (adrianchiris) [reply pkg/sm/plugins/ufm/ufm.go]: Then you will need to deal with old config file (secret) vs new config file (secret)...

File: pkg/sm/plugins/plugin.go (hunk 3 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin.go]: The daemon will read the secret from the api, it won't be mounted to the container.
There is already function the k8s client `GetSecret`

File: pkg/sm/plugins/plugin.go (hunk 4 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin.go]: It should receive parameters of the type that suits its job, no need to overhead it with extracting the guid string from hardware address then remove all ":".

File: pkg/sm/plugins/ufm/ufm.go (hunk 10 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: added support for certificate.

File: pkg/sm/plugins/plugin.go (hunk 4 above)
Comment (adrianchiris) [reply pkg/sm/plugins/plugin.go]: you are defining the API which is two fold
1. for the consumer of an implementation of this API
2. for the implementation itself (for the "plugin" implementor)

From how it is defined now, i dont understand what is the format of guids []string that is passed down to the plugin. (nor the pkey for that matter)

I think you should either use a type that defines the format or have it documented.

Comment (Mmduh-483) [reply pkg/sm/plugins/plugin.go]: It is also possible to add a comment to showing the format, something like:
// pkey: in hexadecimal format e.g. "0x1234"
// guids: array of string guids without ":" in the format e.g. ["001122334455667788"]

Comment (adrianchiris) [reply pkg/sm/plugins/plugin.go]: thats what i said above as an alternative to using types :)

> I think you should either use a type that defines the format or have it documented.

File: pkg/sm/plugins/plugin.go (hunk 3 above)
Comment (moshe010) [reply pkg/sm/plugins/plugin.go]: as @Mmduh-483  said we are getting config map and scripts from kubernetes api no mounting files. So we have already  kubeclient so we don't need to pass the secret as a file.

File: pkg/sm/plugins/plugin.go (hunk 4 above)
Comment (moshe010) [reply pkg/sm/plugins/plugin.go]: I agree with Adrian, we should you strict type so  []net.HardwareAddr makesense, What ufm api get is irrelevant when defining api in the future we nay have plugin which will get it as other type then string. 
Also do we validate pkey value anyware?

File: pkg/sm/plugins/ufm/ufm.go (hunk 13 above)
Comment (moshe010) [reply pkg/sm/plugins/ufm/ufm.go]: @Mmduh-483 , I explain this in the design we should have utils or drivers pkg which contains ways to connect to remote.
for example httpclient /ssh and etc. 
Currently we have on plugin (ufn) which uses one driver (httpclient). we may have other plugins that uses http. you need to decouple the UFM plugin from the driver which using to connect to it.

File: pkg/sm/plugins/ufm/main/main.go [hunk 14]
```diff
@@ -0,0 +1,15 @@
+package main
```
Comment (moshe010) [inline pkg/sm/plugins/ufm/main/main.go]: why we have ufm/main/main.go when we talked about this structure?

File: pkg/sm/plugins/ufm/main/main.go [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+	"github.com/Mellanox/ib-kubernetes/pkg/sm/plugins/ufm"
+
+	"github.com/golang/glog"
+)
+
+var Plugin plugins.SubnetManagerClient
+
+func init() {
+	Plugin = ufm.NewUfmPlugin()
+	glog.Infof("ufm-plugin %v", Plugin)
```
Comment (moshe010) [inline pkg/sm/plugins/ufm/main/main.go]: please init the plugin like in SR-IOV operator [1] and a version to the plugin

[1] - https://github.com/openshift/sriov-network-operator/blob/master/pkg/plugins/mellanox/mellanox_plugin.go#L44-L58

File: pkg/sm/plugins/ufm/ufm.go (hunk 11 above)
Comment (moshe010) [reply pkg/sm/plugins/ufm/ufm.go]: @Mmduh-483 , can you extend init to get the config and change NewUfmPlugin to get config and initialize the client?

File: pkg/sm/plugins/plugin.go (hunk 4 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin.go]: Changed to []net.HardwareAddr and added validation for pkey

File: pkg/sm/plugins/ufm/main/main.go (hunk 14 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/main/main.go]: It would be good if the ufm connection code was in a separate pkg `ufm`  instead of the plugin `main` pkg, since the plugin to be build it should be in main pkg.

File: pkg/sm/plugins/ufm/main/main.go (hunk 15 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/main/main.go]: added

File: pkg/drivers/http/client.go [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+	"net/http"
+
+	"github.com/golang/glog"
+)
+
+type AuthMode string
+
+type Client interface {
+	Get(url string, expectedStatusCode int, body []byte) error
+	Post(url string, expectedStatusCode int, body []byte) error
```
Comment (moshe010) [inline pkg/drivers/http/client.go]: what about Delete? you don't used it?

File: pkg/drivers/http/client.go [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+			caCertPool.AppendCertsFromPEM([]byte(cert))
+			httpClient.Transport.(*http.Transport).TLSClientConfig = &tls.Config{RootCAs: caCertPool}
+		}
+	}
+
+	return &client{authMode: authMode, httpClient: httpClient}
+}
+
+func (c *client) Get(url string, expectedStatusCode int, body []byte) error {
+	glog.V(3).Infof("Get(): url %s, expectedStatusCode %v", url, expectedStatusCode)
```
Comment (moshe010) [inline pkg/drivers/http/client.go]: I would prefer that Get will return the body

File: pkg/ib-utils/utils.go [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+	"strings"
+)
+
+// IsValidatePKey check if the pkey is in the valid range
+func IsValidatePKey(pkey int) bool {
+	return pkey > 0x0000 && pkey < 0xFFFF
+}
+
+// GuidAddrsToStrings return string guids from HardwareAddr
+func GuidAddrsToStrings(addrs []net.HardwareAddr) []string {
```
Comment (moshe010) [inline pkg/ib-utils/utils.go]: change to GuidToString

File: pkg/sm/plugin.go [hunk 19]
```diff
@@ -9,11 +9,11 @@ import (
 	"github.com/golang/glog"
 )
 
-const DefaultPluginSymbolName = "Plugin"
+const DefaultPluginInitializerSymbolName = "Initialize"
```
Comment (moshe010) [inline pkg/sm/plugin.go]: why this change? Is this used just for tests?

File: pkg/sm/plugins/noop/noop.go [hunk 20]
```diff
@@ -1,36 +1,47 @@
...
-	Plugin = plugin{}
-	glog.Infof("noop Plugin: %v %v", Plugin, InvalidPlugin)
+type plugin struct {
+	PluginName  string
+	SpecVersion string
 }
 
-type plugin struct {
+func newNoopPlugin() (*plugin, error) {
+	return &plugin{PluginName: "noop_plugin", SpecVersion: "1.0"}, nil
```
Comment (moshe010) [inline pkg/sm/plugins/noop/noop.go]: rename noop_plugin to noop and move it to const
"1.0" shold move to const

File: pkg/drivers/http/client.go (hunk 16 above)
Comment (Mmduh-483) [reply pkg/drivers/http/client.go]: no, Delete guids is with POST

File: pkg/sm/plugin.go (hunk 19 above)
Comment (Mmduh-483) [reply pkg/sm/plugin.go]: it is need when the daemon loads load the Initialize function from the plugin. it is used in the both tests and the dameon (in the pod event patch will be used)

File: pkg/drivers/http/client.go (hunk 17 above)
Comment (Mmduh-483) [reply pkg/drivers/http/client.go]: Added

File: pkg/ib-utils/utils.go (hunk 18 above)
Comment (Mmduh-483) [reply pkg/ib-utils/utils.go]: Changed

File: pkg/sm/plugins/noop/noop.go (hunk 20 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/noop/noop.go]: Changed

**Comment by Mmduh-483** [comment]
> Why you pushed noop and ufm plugin in the same PR and same commit. It was better to split them to 2PRs or at least 2 commits

Split the patch into 3 commits:
1- update the plugin loader to load initialize function
2- add unit test for noop up plugin
3- implement ufm plugin

File: pkg/sm/plugin.go [hunk 21]
```diff
@@ -9,11 +9,11 @@ import (
...
 
 type PluginLoader interface {
 	// LoadPlugin loads go plugin from given path with given symbolName which is the variable needed to be extracted.
-	LoadPlugin(path, symbolName string) (plugins.SubnetManagerClient, error)
+	LoadPlugin(path, symbolName string) (func(configuration []byte) (plugins.SubnetManagerClient, error), error)
```
Comment (adrianchiris) [inline pkg/sm/plugin.go]: I dont understand what is returned, can you elaborate in the comment ?

Comment (adrianchiris) [reply pkg/sm/plugin.go]: maybe also define a new type alias for func(....) (....)

File: pkg/sm/plugin_test.go [hunk 22]
```diff
@@ -21,17 +21,17 @@ var _ = Describe("Subnet Manager Plugin", func() {
...
 			pl := NewPluginLoader()
-			smClient, err := pl.LoadPlugin(testPlugin, DefaultPluginSymbolName)
+			smClient, err := pl.LoadPlugin(testPlugin, DefaultPluginInitializerSymbolName)
 			Expect(err).ToNot(HaveOccurred())
 			Expect(smClient).ToNot(BeNil())
 		})
 		It("Load non existing plugin", func() {
 			pl := NewPluginLoader()
-			plugin, err := pl.LoadPlugin("not existing", DefaultPluginSymbolName)
+			plugin, err := pl.LoadPlugin("not existing", DefaultPluginInitializerSymbolName)
```
Comment (adrianchiris) [inline pkg/sm/plugin_test.go]: so now it returns a pluginCreate callback right ? maybe rename plugin to : createPluginFunc or something similar throughout the test

File: pkg/sm/plugins/noop/noop.go [hunk 23]
```diff
@@ -1,36 +1,52 @@
...
 
-func (p *plugin) RemovePKey(pkey, guid string) error {
-	log.Println("noop Plugin RemovePKey()")
+func (p *plugin) RemoveGuidsFromPKey(pkey int, guids []net.HardwareAddr) error {
+	glog.V(3).Info("noop Plugin RemovePKey()")
 	return nil
 }
+
+// Initialize applies configs to plugin and return a subnet manager client
+func Initialize(configuration []byte) (plugins.SubnetManagerClient, error) {
```
Comment (adrianchiris) [inline pkg/sm/plugins/noop/noop.go]: you forgot to update this to return pointer to the new func

File: pkg/sm/plugins/plugin.go [hunk 24]
```diff
@@ -1,17 +1,22 @@
 package plugins
 
+import "net"
+
 type SubnetManagerClient interface {
-	// Name returns the subnet manager name.
+	// Name returns the name of the plugin
```
Comment (adrianchiris) [inline pkg/sm/plugins/plugin.go]: nit: name of the client ?

File: pkg/ib-utils/utils.go [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+	"strings"
+)
+
+// IsValidatePKey check if the pkey is in the valid range
+func IsValidatePKey(pkey int) bool {
+	return pkey > 0x0000 && pkey < 0xFFFF
+}
+
+// GuidToString return string guids from HardwareAddr
+func GuidToString(addrs []net.HardwareAddr) []string {
```
[the remark this answers is not in the document]
Comment (moshe010) [reply pkg/ib-utils/utils.go]: I have gave the comment to change it to GudToString, but know I see it take a list of GUIDs.
I prefer to change the method to just to GuidToString and not get all GUIDs as a list.

File: pkg/sm/plugin.go (hunk 19 above)
Comment (moshe010) [reply pkg/sm/plugin.go]: I see so we don't need to call Initialize it will be call automatically?

File: pkg/sm/plugins/ufm/ufm.go [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+
+	httpDriver "github.com/Mellanox/ib-kubernetes/pkg/drivers/http"
+	ibUtils "github.com/Mellanox/ib-kubernetes/pkg/ib-utils"
+	"github.com/Mellanox/ib-kubernetes/pkg/sm/plugins"
+
+	"github.com/golang/glog"
+)
+
+type config struct {
+	Username    string `json:"username"`
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/ufm.go]: AuthMode is missing here.

File: pkg/sm/plugins/ufm/ufm.go [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	return nil
+}
+
+func (u *ufmPlugin) RemoveGuidsFromPKey(pKey int, guids []net.HardwareAddr) error {
+	glog.V(3).Infof("RemoveGuidsFromPKey(): pkey 0x%04X, guids %v", pKey, guids)
+
+	if !ibUtils.IsValidatePKey(pKey) {
+		err := fmt.Errorf("RemoveGuidsFromPKey(): Invalid pkey 0x%04X, out of range 0x0001 - 0xFFFE", pKey)
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/ufm.go]: valid range is 0x0001 - 0x7fff

File: pkg/sm/plugins/ufm/ufm.go [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+		glog.Error(err)
+		return err
+	}
+	guidsString := ibUtils.GuidToString(guids)
+	data := []byte(fmt.Sprintf(`{"pkey": "0x%04X", "index0": true, "ip_over_ib": true, "guids": %q}`,
+		pKey, guidsString))
+
+	if _, err := u.client.Post(u.buildUrl("/ufmRest/resources/pkeys"), http.StatusOK, data); err != nil {
+		err = fmt.Errorf("AddGuidsToPKey(): failed to add guids %v to PKey 0x%04X "+
+			"with error: %v", guids, pKey, err)
```
Comment (adrianchiris) [inline pkg/sm/plugins/ufm/ufm.go]: a general question here : where is the membership of a guid to pkey defined (that is full membership or limited membership)

File: pkg/drivers/http/client.go [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+
+func (c *client) createRequest(method, url string, body io.Reader) (*http.Request, error) {
+	req, err := http.NewRequest(method, url, body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request object %v", err)
+	}
+
+	// Set auth
+	if c.authMode == AuthBasic {
+		req.SetBasicAuth(c.basicAuth.Username, c.basicAuth.Password)
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/drivers/http/client.go]: added

File: pkg/ib-utils/utils.go [hunk 30]
```diff
@@ -0,0 +1,21 @@
+package ib_utils
+
+import (
+	"net"
+	"strings"
+)
+
+// IsValidatePKey check if the pkey is in the valid range
+func IsValidatePKey(pkey int) bool {
+	return pkey > 0x0000 && pkey < 0xFFFF
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/ib-utils/utils.go]: changed to IsPkeyValid.
@adrianchiris the range for the pkey is 0x0001 - 0xFFFE

File: pkg/ib-utils/utils.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/ib-utils/utils.go]: changed to receive one Hardware addr and return the guid string

File: pkg/sm/plugin.go (hunk 21 above)
Comment (Mmduh-483) [reply pkg/sm/plugin.go]: added new type and comments for it.

File: pkg/sm/plugins/noop/noop.go (hunk 23 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/noop/noop.go]: this is the function that will be loaded

File: pkg/sm/plugins/plugin.go (hunk 24 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin.go]: fixed

File: pkg/sm/plugins/ufm/ufm.go (hunk 26 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: It is in the driver/http/client.client

File: pkg/sm/plugins/ufm/ufm.go (hunk 27 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: It's extended to 0x0001 -0xFFFE, already tested it in ufm

File: pkg/sm/plugins/ufm/ufm.go (hunk 28 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: changed to be full always. not sure if we need a config field to be full or limited.

File: pkg/sm/plugin.go (hunk 19 above)
Comment (Mmduh-483) [reply pkg/sm/plugin.go]: no, it will be called in the daemon. the callback initialize is the return value from Loadplugin, then it is going to be called in the daemon.

File: pkg/sm/plugin_test.go (hunk 22 above)
Comment (Mmduh-483) [reply pkg/sm/plugin_test.go]: renamed to CreatePluginFunc

File: pkg/sm/plugins/ufm/ufm.go [hunk 31]
```diff
@@ -0,0 +1,156 @@
...
+type ufmPlugin struct {
+	PluginName  string
+	SpecVersion string
+	conf        *config
+	client      httpDriver.Client
+}
+
+const (
+	PluginName  = "ufm"
+	SpecVersion = "1.0"
```
Comment (moshe010) [inline pkg/sm/plugins/ufm/ufm.go]: we are not going to expose the costs outside of this right? so pluginName and specVersion

File: pkg/sm/plugins/ufm/ufm.go (hunk 28 above)
Comment (moshe010) [reply pkg/sm/plugins/ufm/ufm.go]: I will check what NEO is doing with ib provider and we will do the same

File: pkg/sm/plugins/ufm/ufm.go (hunk 31 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/ufm/ufm.go]: fixed

File: pkg/sm/plugins/noop/noop_test.go [hunk 32]
```diff
@@ -0,0 +1,28 @@
+package main
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Ufm Subnet Manager Client plugin", func() {
```
Comment (moshe010) [inline pkg/sm/plugins/noop/noop_test.go]: you test here the noop plugin not ufm

File: pkg/sm/plugins/ufm/ufm.go (hunk 28 above)
Comment (moshe010) [reply pkg/sm/plugins/ufm/ufm.go]: it ok to keep it as full

File: pkg/sm/plugins/noop/noop_test.go (hunk 32 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/noop/noop_test.go]: fixed
